src/functions.cc

```diff
@@ -23,6 +27,77 @@ NAN_METHOD(set_dev_mode) {
   }
 }
 
+static void addTags(v8::Isolate* isolate, const v8::Local<v8::Object>& object,
+                    Tags* tags) {
+  auto context = isolate->GetCurrentContext();
+  auto maybe_props = object->GetOwnPropertyNames(context);
+  if (!maybe_props.IsEmpty()) {
+    auto props = maybe_props.ToLocalChecked();
+    auto n = props->Length();
+    for (uint32_t i = 0; i < n; ++i) {
+      const auto& key = props->Get(i);
+      const auto& value = object->Get(key);
+      const auto& k = std::string(*Nan::Utf8String(key));
+      const auto& v = std::string(*Nan::Utf8String(value));
+      tags->add(k.c_str(), v.c_str());
+    }
+  }
+}
+
+NAN_METHOD(analyze_id) {
+  const char* err_msg = nullptr;
+  std::string name;
+  ValidationIssues res;
+  Tags tags;
+  auto ret = Nan::New<v8::Array>();
+
+  if (info.Length() > 2) {
+    err_msg =
+        "Expecting at most two arguments: a name and an object describing tags";
+    goto error;
+  }
+
+  if (info.Length() >= 1) {
+    name = *Nan::Utf8String(info[0]);
+    tags.add("name", name.c_str());
+  }
+
+  if (info.Length() == 2 && !(info[1]->IsUndefined() || info[1]->IsNull())) {
+    const auto& maybe_o = info[1];
+    if (maybe_o->IsObject()) {
+      addTags(info.GetIsolate(), maybe_o->ToObject(), &tags);
+    } else {
+      err_msg =
+          "Expected an object with string keys and string values as the second "
+          "argument";
+      goto error;
+    }
+  }
+
+  res = AnalyzeTags(tags);
+  if (res.empty()) {
+    return;
+  }
+
+  for (const auto& issue : res) {
+    auto js_issue = Nan::New<Object>();
+    auto level =
+        issue.level == ValidationIssue::Level::ERROR ? "ERROR" : "WARN";
```

Comment:
What are the semantics of `"WARN"` vs `"ERROR"` issues?

Reply:
An ERROR will cause the metric to be dropped by the backends, so if we're doing validation client side we will just not send it. A WARN means we'll "fix" it for the user (mainly by replacing invalid characters in the name or tags with underscores)

Reply:
OK, it might be worth it to document that somewhere.